hack/generator/spec/defaulter-validator.md

```diff
@@ -0,0 +1,100 @@
+# Custom validation and defaulting for code generated resources
+
+## Reasoning
+`controller-runtime` defines `admission.Defaulter` and `admission.Validator`. These interfaces only give you a single `Default` or `ValidateX` method, which means that all validation/defaulting needs to be done in that method. I think we're going to quickly run into situations where we want custom (handcrafted) validations or defaults for a particular resource and we're not going to want to teach the code generator about these.
+
+## Suggestion
+Structure our autogenerated webhooks such that they're a pipeline. Add the ability to "take control" of this process by implementing an interface.
+
+Implement one (or both) of these interfaces to take control of validation/defaulting by hand
+```
+type Validator interface {
+	CreateValidations() []func() error
+	UpdateValidations() []func(old runtime.Object) error
+	DeleteValidations() []func() error
+}
+
+type Defaulter interface {
+	Defaults() []func()
+}
+```
+
+Sample `Default`:
+```
+// +kubebuilder:webhook:path=/mutate-microsoft-storage-infra-azure-com-v1alpha1api20190401-storageaccount,mutating=true,sideEffects=None,matchPolicy=Exact,failurePolicy=fail,groups=microsoft.storage.infra.azure.com,resources=storageaccounts,verbs=create;update,versions=v1alpha1api20190401,name=default.v1alpha1api20190401.storageaccounts.microsoft.storage.infra.azure.com,admissionReviewVersions=v1beta1
+
+var _ admission.Defaulter = &StorageAccount{}
+
+// Default defaults the Azure name of the resource to the Kubernetes name
+func (storageAccount *StorageAccount) Default() {
+	defaults := storageAccount.defaults()
+	var temp runtime.Object = storageAccount
+	if runtimeDefaulter, ok := temp.(genruntime.Defaulter); ok {
+		defaults = runtimeDefaulter.Defaults()
+	}
+
+	for _, defaultFunc := range defaults {
+		defaultFunc()
+	}
```

Comment:
I didn't click why you needed the cast to `runtime.Object` at first - it's because you can't do type assertions on non-interface values, right? I don't think there is a better way - although since it's nothing specific to `runtime.Object` maybe 
`interface{}` would be a better choice. 

It's quite a weird situation really: "I'm doing an interface check on my own type because I don't know whether I implement this method".

How do you envisage providing custom Defaults implementations?

Reply:
In a handcrafted file alongside the generated code was the assumption. In general I feel like this is a pattern we might want in multiple places (not just here). For example I was imagining @theunrepentantgeek would end up with something similar for conversions eventually for the places we need to deal with manual fixups for them.

Reply:
Changed this to `interface{}`. Also agree that this is weird... is there a less weird way to do this?

Basically I want to encode a hook into the generated code such that we can put handcrafted code "somewhere else" and have it called from here.

Reply:
@theunrepentantgeek actually talks about this same approach here: https://github.com/Azure/k8s-infra/blob/master/hack/generator/spec/versioning.md#standard-extension-points

The only difference is that his proposal has an interface per-type and here we just have a single generic interface, but the same "implement an interface that you don't know about" approach is mentioned.

Reply:
> In a handcrafted file alongside the generated code was the assumption.

Oh, I was worried about something like that going away when we regenerate, but I guess we'll avoid deleting those non-generated files.

Reply:
Yeah we already avoid deleting the non-generated files. I think we look at the `_gen` suffix as well as the first N lines for the comment saying they're generated.